File: exercism/python/phone-number/phone.py

def clean(given):
    chars = [ch for ch in given if ch.isdigit()]
    if not (len(chars) == 10 or len(chars) == 11 and chars[0] == '1'):
        return False
    return ''.join(chars[-10:])


class Phone(object):

    INVALID = '0000000000'

    def __init__(self, given):
        self.number = clean(given) or self.INVALID

    def area_code(self):
        return self.number[:3]

    def pretty(self):
        return '({}) {}-{}'.format(
            self.area_code(),
            self.number[3:6],
            self.number[6:]
        )

File: exercism/haskell/phone-number/Phone.hs
module Phone (areaCode, number, prettyPrint) where

import Data.Char (isDigit)
import Text.Printf (printf)

invalid :: String
invalid = "0000000000"

number :: String -> String
number given
  | length digits == 10 = digits
  | length digits == 11 && head digits == '1' = tail digits
  | otherwise = invalid
  where digits = filter isDigit given


areaCode :: String -> String
areaCode = slice 0 3 . number


prettyPrint :: String -> String
prettyPrint given = printf "(%s) %s-%s" area major minor
  where num = number given
        [area, major, minor] = map ($ num) [areaCode, slice 3 3, slice 6 4]


slice :: Int -> Int -> String -> String
slice a n = take n . drop a
